### PR TITLE
feat(CustomerGroups) Re-write model to allow multiple groups per user

### PR DESCRIPTION
**Pull request by Vince-janv**
## Description
Currently customers can only be part of one customer group. If one customer would like t
### Added

- A link table between customer and customer_groups to create a many-to-many relationship

### How to prepare for test
- [x] ssh to relevant server (depending on type of change)
- [x] Use stage: `us`
- [x] paxa the environment: `paxa`
- [x] install on stage (example for hasta):
`bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-cg-stage.sh modify-customer-groups`
- [x] Create a user that has access to cust004 and cust143

### How to test
- [x] Try creating a MIP order belonging to cust004 and search for a cust003 case
- [x] Try creating a MIP order belonging to cust143 and search for a cust003 case
- [x] Check so that cust004 and cust143 can reorder their own cases
### Expected test outcome
- [x] Check that when choosing cust004 you can select cust003 cases
- [x] Check that when choosing cust143 you cannot select cust003 cases
- [x] Check so that cust143 can select their own case
- [x] Take a screenshot and attach or copy/paste the output.

## Review
- [ ] code approved by
- [x] tests executed by VJ
- [ ] "Merge and deploy" approved by
Thanks for filling in who performed the code review and the test!

### This [version](https://semver.org/) is a
- [ ] **MAJOR** - when you make incompatible API changes
- [x] **MINOR** - when you add functionality in a backwards compatible manner
- [ ] **PATCH** - when you make backwards compatible bug fixes or documentation/instructions

## Implementation Plan
- [ ] Deploy this branch on hasta
- [ ] Deploy branch on cg-prod-services
- [ ] Merge and deploy [cgweb PR](https://github.com/Clinical-Genomics/cgweb/pull/429)
- [ ] Inform to AG

### CODE REVIEW

File: cg/cli/add.py [hunk 1]
```diff
@@ -25,9 +25,8 @@ def add():
 @click.option(
     "-cg",
     "--customer-group",
```
Comment (patrikgrenfeldt) [inline cg/cli/add.py:27]: I think these arguments should get new names, maybe also an automated test?

File: cg/server/admin.py [hunk 2]
```diff
@@ -160,7 +160,7 @@ class CustomerView(BaseView):
         "primary_contact",
         "delivery_contact",
         "priority",
-        "customer_group",
+        "collaborations",
```
Comment (patrikgrenfeldt) [inline cg/server/admin.py:163]: please alphabetize

File: cg/server/api.py [hunk 3]
```diff
@@ -159,17 +159,18 @@ def families():
...
-    """Fetch families in customer_group."""
-    customer_objs: Optional[models.Customer] = (
-        None if g.current_user.is_admin else g.current_user.customers
-    )
-    families_q: Query = db.families_in_customer_group(
-        enquiry=request.args.get("enquiry"), customers=customer_objs
+@BLUEPRINT.route("/families_in_collaboration")
+def families_in_collaboration():
+    """Fetch families in collaboration."""
+    print(request.args.get("customer"))
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:165]: debug?

File: cg/server/api.py [hunk 4]
```diff
@@ -159,17 +159,18 @@ def families():
...
+    order_customer = None if g.current_user.is_admin else db.customer(request.args.get("customer"))
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:166]:
```suggestion
    order_customer = db.customer(request.args.get("customer"))
```

File: cg/server/api.py [hunk 5]
```diff
@@ -159,17 +159,18 @@ def families():
...
+    if not order_customer:
+        return abort(http.HTTPStatus.NOT_FOUND)
+    families_q: Query = db.families(
+        enquiry=request.args.get("enquiry"), customers=order_customer.collaborators
     )
     count = families_q.count()
-    records = families_q.limit(30)
+    records = families_q.limit(1)
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:173]: should be higher than 1?

File: cg/server/api.py [hunk 6]
```diff
@@ -187,16 +188,14 @@ def family(family_id):
     return jsonify(**data)
 
 
-@BLUEPRINT.route("/families_in_customer_group/<family_id>")
-def family_in_customer_group(family_id):
+@BLUEPRINT.route("/families_in_collaboration/<family_id>")
+def family_in_collaboration(family_id):
     """Fetch a family with links."""
     case_obj = db.family(family_id)
+    order_customer = None if g.current_user.is_admin else db.customer(request.args.get("customer"))
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:195]:
```suggestion
    order_customer = db.customer(request.args.get("customer"))
```

File: cg/server/api.py [hunk 7]
```diff
@@ -187,16 +188,14 @@ def family(family_id):
     return jsonify(**data)
 
 
-@BLUEPRINT.route("/families_in_customer_group/<family_id>")
-def family_in_customer_group(family_id):
+@BLUEPRINT.route("/families_in_collaboration/<family_id>")
+def family_in_collaboration(family_id):
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:192]: When is this route used? Can be removed?

File: cg/server/api.py [hunk 8]
```diff
@@ -187,16 +188,14 @@ def family(family_id):
...
     """Fetch a family with links."""
     case_obj = db.family(family_id)
+    order_customer = None if g.current_user.is_admin else db.customer(request.args.get("customer"))
     if case_obj is None:
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:196]:
```suggestion
    if not case_obj or not order_customer:
```

File: cg/server/api.py [hunk 9]
```diff
@@ -250,16 +250,20 @@ def sample(sample_id):
     return jsonify(**data)
 
 
-@BLUEPRINT.route("/samples_in_customer_group/<sample_id>")
-def sample_in_customer_group(sample_id):
+@BLUEPRINT.route("/samples_in_collaboration/<sample_id>")
+def sample_in_collaboration(sample_id):
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:254]: used? can be removed?

File: cg/server/api.py [hunk 10]
```diff
@@ -224,14 +223,15 @@ def samples():
...
-def samples_in_customer_group():
+@BLUEPRINT.route("/samples_in_collaboration")
+def samples_in_collaboration():
     """Fetch samples in a customer group."""
-    customer_objs: Optional[models.Customer] = (
-        None if g.current_user.is_admin else g.current_user.customers
-    )
-    samples_q = db.samples_in_customer_group(
-        enquiry=request.args.get("enquiry"), customers=customer_objs
+    print(request.args.get("customer"))
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:229]: debug?

File: cg/server/api.py [hunk 11]
```diff
@@ -224,14 +223,15 @@ def samples():
...
+    order_customer = None if g.current_user.is_admin else db.customer(request.args.get("customer"))
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:230]:
```suggestion
    order_customer = db.customer(request.args.get("customer"))
```

File: cg/server/api.py [hunk 12]
```diff
@@ -250,16 +250,20 @@ def sample(sample_id):
...
-def sample_in_customer_group(sample_id):
+@BLUEPRINT.route("/samples_in_collaboration/<sample_id>")
+def sample_in_collaboration(sample_id):
     """Fetch a single sample."""
     sample_obj = db.sample(sample_id)
+    order_customer = None if g.current_user.is_admin else db.customer(request.args.get("customer"))
+    print(request.args.get("customer"))
+    if not order_customer:
+        return abort(http.HTTPStatus.NOT_FOUND)
+    print(2)
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:261]: remove

File: cg/server/api.py [hunk 13]
```diff
@@ -250,16 +250,20 @@ def sample(sample_id):
...
     if sample_obj is None:
         return abort(http.HTTPStatus.NOT_FOUND)
-    if not g.current_user.is_admin and (
-        sample_obj.customer.customer_group
-        not in set(customer.customer_group for customer in g.current_user.customers)
-    ):
+    print(3)
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:264]: remove

File: cg/server/api.py [hunk 14]
```diff
@@ -250,16 +250,20 @@ def sample(sample_id):
...
+    if not g.current_user.is_admin and sample_obj.customer not in order_customer.collaborators:
+        print("here!")
```
Comment (patrikgrenfeldt) [inline cg/server/api.py:266]: remove

File: cg/store/api/findbusinessdata.py [hunk 15]
```diff
@@ -175,6 +147,14 @@ def family_samples(self, family_id: str) -> List[models.FamilySample]:
...
         )
 
+    def accessible_customers(self, customer: models.Customer):
+
+        return (
+            self.CustomerLink.query.filter(models.CustomerLink.viewer == customer)
+            .options(load_only(models.CustomerLink.owner))
+            .all()
+        )
+
```
Comment (patrikgrenfeldt) [inline cg/store/api/findbusinessdata.py:157]: What's this method doing?

File: cg/store/models.py [hunk 16]
```diff
@@ -250,17 +254,37 @@ class Customer(Model):
     def __str__(self) -> str:
         return f"{self.internal_id} ({self.name})"
 
-
-class CustomerGroup(Model):
+    @property
+    def collaborators(self) -> Set["Customer"]:
+        """Set of all customers in each customer group in which any of the user's customers are
+        part of"""
```
Comment (patrikgrenfeldt) [inline cg/store/models.py:260]:
```suggestion
        """All customers that the current customer collaborates with (including itself)"""
```

File: tests/cli/add/test_cli_add.py [hunk 17]
```diff
@@ -10,7 +10,7 @@ def test_add_customer(cli_runner: CliRunner, base_context: CGConfig):
     nr_customers: int = status_db.Customer.query.count()
 
     # WHEN adding a customer
-    cli_runner.invoke(
+    result = cli_runner.invoke(
```
Comment (patrikgrenfeldt) [inline tests/cli/add/test_cli_add.py:13]: 👍

File: tests/cli/add/test_cli_add.py [hunk 18]
```diff
@@ -25,22 +25,23 @@ def test_add_customer(cli_runner: CliRunner, base_context: CGConfig):
...
+    assert result.exit_code == 0
     assert status_db.Customer.query.count() == nr_customers + 1
 
 
 def test_add_user(cli_runner: CliRunner, base_context: CGConfig):
     # GIVEN a database with a customer in it that we can connect the user to
     disk_store: Store = base_context.status_db
     customer_id = "custtest"
-    customer_group = disk_store.add_customer_group("dummy_group", "dummy group")
+    collaboration = disk_store.add_collaboration("dummy_group", "dummy group")
```
Comment (patrikgrenfeldt) [inline tests/cli/add/test_cli_add.py:36]:
```suggestion
```

File: tests/cli/add/test_cli_add.py [hunk 19]
```diff
@@ -25,22 +25,23 @@ def test_add_customer(cli_runner: CliRunner, base_context: CGConfig):
...
     customer = disk_store.add_customer(
         internal_id=customer_id,
         name="Test Customer",
         scout_access=False,
-        customer_group=customer_group,
         invoice_address="Street nr, 12345 Uppsala",
         invoice_reference="ABCDEF",
     )
+    customer.collaborations.append(collaboration)
```
Comment (patrikgrenfeldt) [inline tests/cli/add/test_cli_add.py:44]:
```suggestion
```

File: tests/store_helpers.py [hunk 20]
```diff
@@ -155,13 +155,13 @@ def ensure_customer(
         customer_id: str = "cust000",
         name: str = "Production",
         scout_access: bool = False,
-        customer_group: str = "all_customers",
+        collaboration: str = "all_customers",
```
Comment (patrikgrenfeldt) [inline tests/store_helpers.py:158]:
```suggestion
        collaboration_id: str = "all_customers",

```

File: tests/store_helpers.py [hunk 21]
```diff
@@ -155,13 +155,13 @@ def ensure_customer(
...
         scout_access: bool = False,
-        customer_group: str = "all_customers",
+        collaboration: str = "all_customers",
     ) -> models.Customer:
         """Utility function to return existing or create customer for tests"""
-        customer_group_id = customer_group or customer_id + "_group"
-        customer_group = store.customer_group(customer_group_id)
-        if not customer_group:
-            customer_group = store.add_customer_group(customer_group_id, customer_group_id)
+        collaboration_id = collaboration or f"{customer_id}_group"
```
Comment (patrikgrenfeldt) [inline tests/store_helpers.py:161]: this seems a bit redundant since we have a default name already on line 158

**Review by patrikgrenfeldt — COMMENTED** [review]
I think we should have a look on the comments together and discuss our opinions 😄

File: alembic/versions/ddc94088be4d_adds_customer_to_customer_group_link_.py [hunk 22]
```diff
@@ -0,0 +1,124 @@
...
+        sa.ForeignKeyConstraint(
+            ("collaboration_id",),
+            ["collaboration.id"],
+        ),
+        sa.UniqueConstraint("customer_id", "collaboration_id", name="_customer_collaboration_uc"),
+    )
+    bind = op.get_bind()
+    session = sa.orm.Session(bind=bind)
+    for collaboration in session.query(Collaboration):
+        print(f"Customer group {collaboration.internal_id} contains:")
```
Comment (Vince-janv) [inline alembic/versions/ddc94088be4d_adds_customer_to_customer_group_link_.py:70]: Too much text!

File: cg/server/api.py (hunk 7 above)
Comment (Vince-janv) [reply cg/server/api.py:192]: It is used by cgweb

File: cg/server/api.py (hunk 9 above)
Comment (Vince-janv) [reply cg/server/api.py:254]: It is used by cgweb

**Review by patrikgrenfeldt — APPROVED** [review]
👍

**Pull request merged**